### PR TITLE
Feat/home 53 color picker and limitations

### PR DESCRIPTION
**Pull request by Phabes**

### CODE REVIEW

File: views/FanView/FanView.tsx [hunk 1]
```diff
@@ -16,6 +16,8 @@ import { styles } from './FanView.styles';
 
 type FanViewProps = NativeStackScreenProps<RootStackParamList, 'Fan'>;
 
+const numberRegex = new RegExp('\\.', 'g');
```
Comment (mateusz-kleszcz) [inline views/FanView/FanView.tsx:19]: Do we need this regex? I think that setting `keyboardType` as `numeric` is sufficient security

File: views/FanView/FanView.tsx [hunk 2]
```diff
@@ -56,15 +61,60 @@ export const FanView: FC<FanViewProps> = ({ route }) => {
...
+    } else {
+      numberOfDots = value.match(numberRegex)?.length;
+      const finalValueString =
+        value == finalValueFloat.toString()
+          ? finalValueFloat.toString()
+          : numberOfDots && numberOfDots > 0
+          ? finalValueFloat.toString() + '.'
+          : finalValueFloat.toString();
+      inputChange(finalValueString);
+    }
```
Comment (mateusz-kleszcz) [inline views/FanView/FanView.tsx:115]: This code is really complex and hard to read. You can simplify it like this
```
    const newValue = value.replace(',', '.').split('.').slice(0, 2).join('.');
    const newValueNumeric = parseFloat(newValue);
    if (newValueNumeric < min) {
      inputChange(min.toString());
    } else if (newValueNumeric > max) {
      inputChange(max.toString());
    } else {
      inputChange(newValue);
    }
```
Maybe you can also use `Math.min` and `Math.max` to simplify this `if` statements, but I've got a problem with a comma. But you can give it a try :D

File: views/FanView/FanView.tsx [hunk 3]
```diff
@@ -74,31 +124,57 @@ export const FanView: FC<FanViewProps> = ({ route }) => {
...
+        <View style={styles.content}>
+          <View style={styles.section}>
+            <Typography
+              variant="body-medium"
+              text={`Aktualna temperatura wentylacji: ${fan.values.V_TEMP}°C`}
+            />
+            <Input
+              text={temp}
+              onChange={e => {
+                checkInput(e, onTempChange, 0, 50);
```
Comment (mateusz-kleszcz) [inline views/FanView/FanView.tsx:137]: Instead of passing `checkInput` as a onChange function, move the code that is checking if input is correct and run it automatically if `keyboardType` is equal to numeric. Also, a min and a max should be a prop (similar to input type="numeric" in HTML)

File: views/RgbLightView/RgbLightView.tsx [hunk 4]
```diff
@@ -42,14 +46,16 @@ export const RgbLightView: FC<RgbLightViewProps> = ({ route }) => {
...
       <Navbar text={`${location} - ${nodeId} - ${rgbLight?.name}`} />
-      <View style={styles.content}>
-        <Typography
-          variant="body-medium"
-          text={`Aktualny kolor lampy to: ${rgbLight.values.V_RGB}`}
-        />
-        <Input text={color} onChange={onColorChange} />
-        <Button text="Zmień kolor" onPress={handleChangeColor} />
-      </View>
+      <ScrollView onScrollBeginDrag={() => Keyboard.dismiss()}>
```
Comment (mateusz-kleszcz) [inline views/RgbLightView/RgbLightView.tsx:49]:
```suggestion
      <ScrollView onScrollBeginDrag={Keyboard.dismiss}>
```
Is it working? I didn't test it but seems fine

File: views/RgbLightView/RgbLightView.tsx [hunk 5]
```diff
@@ -23,7 +22,12 @@ export const RgbLightView: FC<RgbLightViewProps> = ({ route }) => {
     selectDeviceWithId(state, location, nodeId, deviceId),
   ) as RgbLight;
 
-  const [color, onColorChange] = useInputValue(rgbLight.values.V_RGB);
+  const [color, onColorChange] = useState(rgbLight.values.V_RGB);
+
+  const colorPickerColorChange = (rgbColor: string) => {
+    const colorValue = rgbColor.split('#')[1];
```
Comment (mateusz-kleszcz) [inline views/RgbLightView/RgbLightView.tsx:28]: Why color is with hash? Just return the raw value in the RGB picker component

**Review by mateusz-kleszcz — CHANGES_REQUESTED** [review]

File: .storybook/stories/Input/Input.tsx [hunk 6]
```diff
@@ -8,8 +13,11 @@ export type InputProps = {
   placeholder?: string;
   variant?: 'default' | 'error';
   disabled?: boolean;
+  autoCapitalize?: 'none' | 'sentences' | 'words' | 'characters';
   centerText?: boolean;
   onChange: (e: string) => void;
+  onBlur?: (e: string) => void;
```
[the remark this answers is not in the document]
Comment (Phabes) [reply .storybook/stories/Input/Input.tsx:19]: Idk if it is what you expected

[the remark this answers is not in the document]
Comment (mateusz-kleszcz) [reply .storybook/stories/Input/Input.tsx:19]: I mean, could you change these handle event functions so they look like this:
```
onSomething?: (value?: string) => void;
```

[the remark this answers is not in the document]
Comment (Phabes) [reply .storybook/stories/Input/Input.tsx:19]: Didn't commit changes yet. But it is what I've done

File: views/RgbLightView/RgbLightView.tsx (hunk 5 above)
Comment (Phabes) [reply views/RgbLightView/RgbLightView.tsx:28]: Should for fine

File: views/FanView/FanView.tsx (hunk 3 above)
Comment (Phabes) [reply views/FanView/FanView.tsx]: Done

File: .storybook/stories/Input/Input.tsx [hunk 7]
```diff
@@ -18,24 +28,65 @@ export const Input: FC<InputProps> = ({
...
+      setType(variant);
+      onBlur?.(value.nativeEvent.text);
+    },
+    [variant],
+  );
+
+  const getNumberValue = (value: string): string => {
+    if (value === '') {
+      value = min.toString();
+    }
```
Comment (mateusz-kleszcz) [inline .storybook/stories/Input/Input.tsx:63]: 👍  gj

File: .storybook/stories/Input/Input.tsx [hunk 8]
```diff
@@ -18,24 +28,65 @@ export const Input: FC<InputProps> = ({
...
+    const parts = value.replace(',', '.').split('.').slice(0, 2);
+    parts[0] = parseInt(parts[0]).toString();
```
Comment (mateusz-kleszcz) [inline .storybook/stories/Input/Input.tsx:65]: What exactly is happening in this line? Why you are parsing the first part of the value and then casting it to a string again? It doesn't make much sense to me. If you add a `.join('.')` to a previous line, it should work correctly

File: .storybook/stories/Input/Input.tsx [hunk 9]
```diff
@@ -18,24 +28,65 @@ export const Input: FC<InputProps> = ({
...
+    const newValue = parts.join('.');
+    const newValueNumeric = parseFloat(newValue);
+    const finalValue =
+      newValueNumeric < min
+        ? min.toString()
+        : newValueNumeric > max
+        ? max.toString()
+        : newValue;
```
Comment (mateusz-kleszcz) [inline .storybook/stories/Input/Input.tsx:73]: Don't use nested ternary operators. Just write
```
if (newValueNumeric < min) {
return min.toString()
} else if (newValueNumeric > max) {
return max.toString()
}
return newValueNumeric
```
This takes the same number of lines, but it's much more readable and easy to understand.

File: .storybook/stories/Input/Input.tsx [hunk 10]
```diff
@@ -67,7 +118,6 @@ const useStyles = (
       color: theme.colors[textColor],
       borderColor: theme.colors[borderColor],
       width: '100%',
-      ...theme.typography['body-medium'],
```
Comment (mateusz-kleszcz) [inline .storybook/stories/Input/Input.tsx:70]: Is there any specific reason why you removed typography from useStyles hook and moved it to the style object inside a component? I think here it's a better place

File: .storybook/stories/Typography/Typography.tsx [hunk 11]
```diff
@@ -15,10 +15,11 @@ export const Typography: FC<TypographyProps> = ({
...
 }) => {
-  const styles = StyleSheet.create({
-    ...theme.typography[variant],
-    color: theme.colors[color],
-  });
-
-  return <Text style={styles} selectable={selectable}>{text}</Text>;
+  return (
+    <Text
+      style={[{ color: theme.colors[color] }, theme.typography[variant]]}
```
Comment (mateusz-kleszcz) [inline .storybook/stories/Typography/Typography.tsx:20]: Same here

File: views/FanView/FanView.tsx [hunk 12]
```diff
@@ -24,26 +33,29 @@ export const FanView: FC<FanViewProps> = ({ route }) => {
...
-  const [direction, onDirectionChange] = useInputValue(fan.values.V_DIRECTION);
   const [percentage, onPercentageChange] = useInputValue(
     fan.values.V_PERCENTAGE,
   );
+  const [direction, onDirectionChange] = useInputValue(fan.values.V_DIRECTION);
 
   const handleChangeFanParams = () => {
+    const finalTempValue = parseFloat(temp).toString();
+    const finalDirectionValue = parseFloat(direction).toString();
+    const finalPercentageValue = parseFloat(percentage).toString();
```
Comment (mateusz-kleszcz) [inline views/FanView/FanView.tsx:44]: Why?

File: views/FanView/FanView.tsx [hunk 13]
```diff
@@ -10,12 +10,21 @@ import {
...
 import { selectDeviceWithId } from '../../redux/devicesSlice';
 import { useAppSelector } from '../../redux/hooks';
-import { Fan, RootStackParamList } from '../../types';
+import { Fan, FanRangeValues, RootStackParamList } from '../../types';
 import { sendAPIRequest } from '../../utils';
 import { styles } from './FanView.styles';
 
 type FanViewProps = NativeStackScreenProps<RootStackParamList, 'Fan'>;
 
+const fanRangeValues: FanRangeValues = {
```
Comment (mateusz-kleszcz) [inline views/FanView/FanView.tsx:19]: Good idea to move it to const :D however, the problem with a const object is that in the JS const isn't a constant, it only means that the variable cannot be re-assigned. It means that you cannot change the whole object, but you can change the value of a particular properties
image
I would consider using a enum or moving it to a separate const (remember to use the correct letter case - for enum PascalCase, for const CONST_CASE)

File: views/RgbLightView/RgbLightView.tsx [hunk 14]
```diff
@@ -42,14 +41,16 @@ export const RgbLightView: FC<RgbLightViewProps> = ({ route }) => {
...
       <Navbar text={`${location} - ${nodeId} - ${rgbLight?.name}`} />
-      <View style={styles.content}>
-        <Typography
-          variant="body-medium"
-          text={`Aktualny kolor lampy to: ${rgbLight.values.V_RGB}`}
-        />
-        <Input text={color} onChange={onColorChange} />
-        <Button text="Zmień kolor" onPress={handleChangeColor} />
-      </View>
+      <ScrollView onScrollBeginDrag={Keyboard.dismiss}>
```
Comment (mateusz-kleszcz) [inline views/RgbLightView/RgbLightView.tsx:44]: In my last PR, I moved this ScrollView to a Layout component. You can consider moving onScrollBeginDraw props there

**Review by mateusz-kleszcz — CHANGES_REQUESTED** [review]
Please resolve conflicts first

File: views/FanView/FanView.tsx [hunk 15]
```diff
@@ -74,31 +86,69 @@ export const FanView: FC<FanViewProps> = ({ route }) => {
...
+              text={`Aktualny kierunek głowicy: ${fan.values.V_DIRECTION}°`}
+            />
+            <Typography
+              variant="body-small"
+              text={`Zakres: ${fanRangeValues.minDirection} - ${fanRangeValues.maxDirection}`}
+            />
+            <Input
+              text={direction}
+              onChange={onDirectionChange}
+              placeholder=""
```
Comment (mateusz-kleszcz) [inline views/FanView/FanView.tsx:137]: One more, why did you add an empty placeholder? Just removed it in all places :D

File: .storybook/stories/Input/Input.tsx (hunk 10 above)
Comment (Phabes) [reply .storybook/stories/Input/Input.tsx:70]: Because it's causing errors. Now everything seems fine and vscode doesn't show errors.

File: views/FanView/FanView.tsx (hunk 12 above)
Comment (Phabes) [reply views/FanView/FanView.tsx:44]: To avoid situation like when in input is something like `54.`

File: .storybook/stories/Input/Input.tsx (hunk 8 above)
Comment (Phabes) [reply .storybook/stories/Input/Input.tsx:65]: Leading `0` are deleting there.

File: views/FanView/FanView.tsx (hunk 15 above)
Comment (Phabes) [reply views/FanView/FanView.tsx:137]: Will change placeholder prop to empty initially

File: views/FanView/FanView.tsx (hunk 13 above)
Comment (Phabes) [reply views/FanView/FanView.tsx:20]: I know about that. Changes applied.

File: views/RgbLightView/RgbLightView.tsx (hunk 14 above)
Comment (Phabes) [reply views/RgbLightView/RgbLightView.tsx]: Done but not in a prop but just inside scroll view.

File: .storybook/stories/ColorPickerRGB/ColorPickerRGB.tsx [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+        placeholder="Podaj kolor RGB"
+      />
+    </View>
+  );
+};
+
+const styles = StyleSheet.create({
+  container: {
+    flex: 1,
+    flexDirection: 'column',
```
Comment (mateusz-kleszcz) [inline .storybook/stories/ColorPickerRGB/ColorPickerRGB.tsx:58]: Isn't it default?

**Review by mateusz-kleszcz — APPROVED** [review]
